hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImplV2.java

```diff
@@ -0,0 +1,412 @@
+            // Mostly likely it's a retried delete command response.
+            if (LOG.isDebugEnabled()) {
+              LOG.debug(
+                  "Transaction txId={} commit by dnId={} for containerID={}"
+                      + " failed. Corresponding entry not found.", txID, dnID,
+                  containerId);
+            }
+            continue;
+          }
+
+          dnsWithCommittedTxn.add(dnID);
+          final ContainerInfo container =
+              containerManager.getContainer(containerId);
+          final Set<ContainerReplica> replicas =
+              containerManager.getContainerReplicas(containerId);
+          // The delete entry can be safely removed from the log if all the
+          // corresponding nodes commit the txn. It is required to check that
+          // the nodes returned in the pipeline match the replication factor.
+          if (min(replicas.size(), dnsWithCommittedTxn.size())
+              >= container.getReplicationFactor().getNumber()) {
+            List<UUID> containerDns = replicas.stream()
+                .map(ContainerReplica::getDatanodeDetails)
+                .map(DatanodeDetails::getUuid)
+                .collect(Collectors.toList());
+            if (dnsWithCommittedTxn.containsAll(containerDns)) {
+              transactionToDNsCommitMap.remove(txID);
+              if (LOG.isDebugEnabled()) {
+                LOG.debug("Purging txId={} from block deletion log", txID);
+              }
+              transactionIDs.addTxID(txID);
+            }
+          }
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("Datanode txId={} containerId={} committed by dnId={}",
+                txID, containerId, dnID);
+          }
+        } catch (IOException e) {
+          LOG.warn("Could not commit delete block transaction: " +
+              transactionResult.getTxID(), e);
+        }
+      }
+      try {
+        deletedBlockLogStateManager
+            .removeTransactionsFromDB(transactionIDs.build());
+      } catch (IOException e) {
+        LOG.warn("Could not commit delete block transactions: " +
+            transactionIDs.build().getTxIDList(), e);
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private boolean isTransactionFailed(DeleteBlockTransactionResult result) {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug(
+          "Got block deletion ACK from datanode, TXIDs={}, " + "success={}",
+          result.getTxID(), result.getSuccess());
+    }
+    if (!result.getSuccess()) {
+      LOG.warn("Got failed ACK for TXID={}, prepare to resend the "
+          + "TX in next interval", result.getTxID());
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @param containerID - container ID.
+   * @param blocks      - blocks that belong to the same container.
+   * @throws IOException
+   */
+  @Override
+  public void addTransaction(long containerID, List<Long> blocks)
```

Comment:
We will not support this method `addTransaction()` in HA code, since it is only used in test code, we need change affected tests as well.